kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/internal/queue/PaymentQueueManagerImpl.java

```diff
@@ -0,0 +1,112 @@
+        this.pendingBalanceUpdater = pendingBalanceUpdater;
+        this.eventsManager = eventsManager;
+        this.delayBetweenPayments = delayBetweenPayments;
+        this.queueTimeout = queueTimeout;
+        this.maxNumOfPayments = maxNumOfPayments;
+        queue = new ArrayList<>(maxNumOfPayments);
+        delayTask = new DelayTask();
+    }
+
+    @Override
+    public void enqueue(final PendingPayment pendingPayment) {
+        queueScheduler.schedule(new Runnable() {
+            @Override
+            public void run() {
+                // 1. validate pending balance
+                validatePendingBalance(); // TODO: 2019-08-15 handle it
+
+                if (getPendingPaymentCount() == maxNumOfPayments - 1) {
+                    // 2. add to queue
+                    // 3. cancel scheduling
+                    // 4. maybe events in later stages
+                    // 5. clear and send to task queue manager
+
+                    queue.add(pendingPayment);
+                    // TODO: 2019-08-08 maybe send events here
+                    sendPendingPaymentsToTaskQueue();
+                } else {
+                    addToQueue();
+                }
+            }
+
+            private void addToQueue() {
+                // 1. add to list
+                // 2. handle events later (maybe also change the execution)
+                // 3. reset and schedule delay
+                // 4. schedule timeout if first in list
+
+                queue.add(pendingPayment);
+                // TODO: 2019-08-08 maybe send events here
+                queueScheduler.removePendingTask(delayTask);
+                delayTask = new DelayTask();
+                queueScheduler.scheduleDelayed(delayTask, delayBetweenPayments);
+                if (queue.size() == 1) {
+                    queueScheduler.scheduleDelayed(new DelayTask(), queueTimeout);
+                }
+            }
+        });
+    }
+
+    @Override
+    public List<PendingPayment> getPaymentQueue() {
+        // TODO: 2019-08-15 need to think why exactly if we really need it? is it only for the pending balance updater?
+        //  maybe return a copy or make it immutable/read only...
+        return queue;
+    }
+
+    @Override
+    public int getPendingPaymentCount() {
+        return queue.size();
+    }
+
+    private void validatePendingBalance() {
+        Balance pendingBalance = pendingBalanceUpdater.getPendingBalance();
+        // TODO: 2019-08-08 implement check and validate the pending balance
+    }
+
+    private void sendPendingPaymentsToTaskQueue() {
+        queueScheduler.removeAllPendingTasks();
+        // TODO: 2019-08-08 maybe send events
+        List<PendingPayment> pendingPayments = queue;
+        queue = new ArrayList<>(maxNumOfPayments);
+        txTaskQueueManager.enqueue(pendingPayments);
+    }
+
+    private class DelayTask implements Runnable {
+
+        @Override
+        public void run() {
+            validatePendingBalance();
```

Comment:
But maybe when we actually want to send it then there is not enough balance?
or we don't care for this?

Reply:
good point, actually I think the correct solution is to check the pending balance also with transaction params, this will make sure that we never enter to the queue any tx/payment that will go below our balance. FYI @cnotethegr8